sycl/source/detail/program_manager/program_manager.hpp

```diff
@@ -284,6 +285,13 @@ class ProgramManager {
   /// \ref Sync::getGlobalLock() while holding this mutex.
   std::mutex m_KernelIDsMutex;
 
+  /// Caches all found service kernels to expedite future checks. A SYCL service
+  /// kernel is a kernel that has not been defined by the user but is instead
+  /// generated by the SYCL runtime. Service kernel name types must be declared
+  /// in the sycl::detail::__sycl_service_kernel__ namespace which is
+  /// exclusively used for this purpose.
+  std::unordered_set<std::string> m_ServiceKernels;
```

Comment:
Probably, a note should be placed here saying that write access is only allowed during start-up process in `addImages`.

Reply:
I have added such a note.

Reply:
Sorry for complicating things, but I'm not sure we can rely on the fact that `addImages` can be called only during process startup. A user can `dlopen` a library which contains SYCL kernel and that will be registered by calling `addImages` way after process startup.

Reply:
That is a good point. That said, I don't think we need a new mutex as the kernel ID mutex will protect it. It does mean that we can't remove the Kernel ID mutex though. I'll close that PR and change this note to mention the kernel ID mutex instead..